agimus_controller/agimus_controller/robot_model_factory/robot_model_factory.py

```diff
@@ -0,0 +1,144 @@
+    _collision_parser = ObstacleParamsParser()
+    """ Paramters of the model. """
+    _params = RobotModelParameters()
+    """ Path to the collisions environment. """
+    _env = Path()
+
+    @classmethod
+    def load_model(cls, param: RobotModelParameters, env: Union[Path, None]) -> Self:
+        model = cls()
+        model._params = param
+        model._env = env
+        model._load_pinocchio_models(param.urdf, param.free_flyer)
+        model._load_default_configuration(param.srdf, param.q0_name)
+        model._load_reduced_model(param.locked_joint_names, param.q0_name)
+        model._update_collision_model(
+            env, param.collision_as_capsule, param.self_collision, param.srdf
+        )
+        return model
+
+    def _load_pinocchio_models(self, urdf: Path, free_flyer: bool) -> None:
+        verbose = False
+        assert urdf.exists() and urdf.is_file()
+        if free_flyer:
+            self._model, self._cmodel, self._vmodel = pin.buildModelsFromUrdf(
+                filename=str(urdf),
+                root_joint=pin.JointModelFreeFlyer(),
+                verbose=verbose,
+            )
+        else:
+            self._model, self._cmodel, self._vmodel = pin.buildModelsFromUrdf(
+                filename=str(urdf), root_joint=None, verbose=verbose
+            )
+
+    def _load_default_configuration(self, srdf_path: Path, q0_name: str) -> None:
+        if not srdf_path.is_file():
+            return
+        pin.loadReferenceConfigurations(self._model, str(srdf_path), False)
+        if q0_name and q0_name in self._model.referenceConfigurations:
+            self._q0 = self._model.referenceConfigurations[q0_name]
+        else:
+            self._q0 = pin.neutral(self._model)
+
+    def _load_reduced_model(self, locked_joint_names, q0_name) -> None:
+        locked_joint_ids = [self._model.getJointId(name) for name in locked_joint_names]
+        self._rmodel, geometric_models_reduced = pin.buildReducedModel(
+            self._model,
+            list_of_geom_models=[self._cmodel, self._vmodel],
+            list_of_joints_to_lock=locked_joint_ids,
+            reference_configuration=self._q0,
+        )
+        self._rcmodel, self._rvmodel = geometric_models_reduced
+
+        if q0_name and q0_name in self._rmodel.referenceConfigurations:
+            self._q0 = self._rmodel.referenceConfigurations[q0_name]
+        else:
+            self._q0 = pin.neutral(self._rmodel)
+
+    def _update_collision_model(
+        self,
+        env: Union[Path, None],
+        collision_as_capsule: bool,
+        self_collision: bool,
+        srdf: Path,
+    ) -> None:
+        if collision_as_capsule:
+            self._rcmodel = self._collision_parser.transform_model_into_capsules(
+                self._rcmodel
+            )
+        if self_collision and srdf.exists():
+            self._rcmodel = self._collision_parser.add_self_collision(
+                self._rmodel, self._rcmodel, srdf
+            )
```

Comment:
This one is dangerous and needs some double check when using automatic self collisions, I'd maybe put a warning saying to verify the collision pairs afterward?

Reply:
Ok what did you see as issues before? Too many pairs?

Reply:
Yep